File: Problems/322/main.go
package main

import "fmt"

func main() {

	coinChange := coinChange

	fmt.Println(coinChange([]int{1, 2, 5}, 11))
	fmt.Println(coinChange([]int{2}, 3))
	fmt.Println(coinChange([]int{1}, 0))
}

func coinChange(coins []int, amount int) int {
	// DP Bottom UP
	// O(n)
	// with no improvement
	bc := make([]int, amount+1)
	for i := range bc {
		bc[i] = -1
	}
	bc[0] = 0

	for i := 1; i <= amount; i++ {
		counts := 100000
		for _, c := range coins {
			t := i - c
			if t >= 0 && bc[t] != -1 && bc[t]+1 < counts {
				counts = bc[t] + 1
			}
		}
		if counts != 100000 {
			bc[i] = counts
		}
	}

	return bc[amount]
}

File: Problems/322/main.cc
#include <vector>
#include <iostream>
#include <string>
#include <iomanip>

using namespace std;

class Solution
{
public:
    int coinChange(vector<int> &coins, int amount)
    {
        // Initial Backtracking Solution
        // Generate every combination and get the minimum count of combination

        // So Let's change to 2D DP solution to avoid redundant calculation
        // i -> coins kind
        // j -> amount
        // dp[i][j] = min(dp[i-1][j], dp[i][j-coins[i]]+1)

        int n = coins.size();
        int m = amount;
        int dp[n + 1][m + 1];

        for (int j = 0; j <= m; j++)
            dp[0][j] = -1; // Meaning Can't be Done

        for (int i = 0; i <= n; i++)
            dp[i][0] = 0; // Automatically Done

        for (int i = 1; i <= n; i++)
        {
            for (int j = 1; j <= m; j++)
            {
                dp[i][j] = min(
                    dp[i - 1][j] == -1 ? INT_MAX : dp[i - 1][j],
                    j >= coins[i - 1] && dp[i][j - coins[i - 1]] != -1 ? dp[i][j - coins[i - 1]] + 1 : INT_MAX);
                if (dp[i][j] == INT_MAX)
                    dp[i][j] = -1;
            }
        }

        for (int i = 0; i <= n; i++)
        {
            for (int j = 0; j <= m; j++)
            {
                cout << setw(4) << dp[i][j];
            }
            cout << endl;
        }

        return dp[n][m];
    }
};

int main()
{
    vector<int> coins{1, 2, 5};
    Solution sol;
    sol.coinChange(coins, 11);

    return 0;
}
